### PR TITLE
Qwen3.6-35B-A3B support: staged engine + N-stage sharded pipeline

### PR DESCRIPTION
**Pull request by pawanpaudel93**
Part B of #77 on the community side: full Qwen3.6-35B-A3B (`qwen3_5_moe`) support — nonsharded serving + the staged sharded engine, **single-box and the M4'-1 cross-node N-stage pipeline**. Base branch is the 2026.2 upgrade (Part A).

## What's in here

**Staged engine (`--engine qwen36-moe`)**
- IR-surgery shard exporter (`tools/qwen36_surgery/export_qwen36_moe.py`): cuts the official int4 OV IR at decoder-layer boundaries; per-layer DeltaNet conv/ssm + attention KV state ride along as ReadValue/Assign; orphan-state + ShapeOf rewires; last-stage logit slice; 8-token greedy validation in `--validate`; copies the aux files (tokenizer / chat_template / text-embeddings) into the tree alongside `manifest.json` + `stage{i}/`.
- `Qwen36Builder`: in-process stage chain (`--total 1`) **and the M4'-1 N-stage cross-node pipeline** (`with_rank`, `configure_listen`, `connect` / `forward_downstream` / `step_pipe_relay`). Rank 0 holds embeddings + stage 0 + tokenizer and drives decode; middle ranks relay (run their stage, forward the span downstream, return the token); the last rank holds the logits head and answers each FORWARD with the argmax. 12-byte BE frame header `[kind][epoch][pos]`, HELLO/RESET/FORWARD/TOKEN, lockstep per hop, stale-epoch drop. Incremental `step()` (per-token streaming + working cancel), chunked batched prefill.
- `cascadia shard` dispatch arm for `qwen3_5_moe` (same-command UX; torch lazy — the IR-surgery path runs on openvino + numpy alone).

**API/runtime fixes that fell out**
- real `prompt_tokens` in usage (Chunk carries it on the final marker)
- `enable_thinking` request flag; staged engine honors `false`
- root-layout chat-template loader (`load_chat_template_config_at`) — reads `chat_template.jinja` at the model root (the surgery tree has no `tokenizer/` subdir)

## Validation
- **Single-box**: engine ≡ Python chain 64/64 greedy; golden regression test; E2E API 5/5; cancel/disconnect robustness; per-token streaming; fixed prompt set 6/6 on both paths; well-formed hermes `<tool_call>`.
- **Cross-node N-stage** (cascadia-pawan fleet, OV 2026.2, via the enterprise embedded integration cascadia-enterprise#36): 2/3/4-stage chains form / handshake / serve on real hardware. **2-stage and 3-stage validated clean** (constrained Q&A, burst, and sustained soak all correct); **4-stage forms and serves but output drifts** — see limitations.
- **iGPU (device=GPU)**: on the 3-stage GPU chain — 6-prompt set 6/6 (Paris/Tokyo/7x8/mice/Moo/Nile), hermes `<tool_call>` well-formed, 10-way burst 10/10, splice CONFIRMED (GREEN).
- **ov-genai `enable_thinking`**: validated before/after on pawan-01 (CPU) — `enable_thinking:false` returns a direct answer (no `<think>`), `true` reasons. Confirmed on **Qwen3-1.7B-int4-ov** (LLMPipeline) and **Qwen3.6-35B-A3B-int4-ov** (VLMPipeline). **Hybrid**: thinking-OFF renders the model's template API-side (root- or `tokenizer/`-layout) + `apply_chat_template=false`; thinking-ON stays on ov-genai's **native** template, untouched — so the already-working path is byte-identical and only the broken path changes.
- chain-vs-single-box greedy is prompt-dependent — near-ties + cross-GPU float variance flip some tokens; the engine is **bit-faithful to its own chain** (head decode verified bit-exact via an activation-parity trace). See `docs/architectures/qwen36-moe-support.md` + the M4' pipeline spec.

## Test plan (reviewer)

**Run every command in PowerShell on the Windows box.** Each is a single line. **A** = CI gate + shard export. **B** = OpenVINO serving (needs the 2026.2 GenAI SDK). **C** = the Windows fleet, owned by cascadia-enterprise#36.

### A — CI gate + export

Get the branch + reproduce CI (`ci.yml`; no OpenVINO, no model):

```powershell
cargo fmt --all -- --check
cargo clippy --workspace --all-targets
cargo test --workspace --all-targets
```

Export a 2-stage shard tree (OpenVINO python wheel + numpy):

```powershell
pip install -U "openvino>=2026.2" numpy huggingface_hub
hf download OpenVINO/Qwen3.6-35B-A3B-int4-ov --local-dir ./qwen36-int4-ov
python tools/qwen36_surgery/export_qwen36_moe.py --model ./qwen36-int4-ov --out ./qwen36-shards-2stage --total 2 --validate
```

### B — OpenVINO serving (Intel GPU/NPU; CPU works but slow)

OpenVINO is needed at **build time** (`INTEL_OPENVINO_DIR`) **and run time** (its DLLs on `PATH`). `setupvars.ps1` sets the build var; then prepend the runtime DLL dirs (else the test/serve exe fails with `STATUS_DLL_NOT_FOUND`). The rig installs the SDK at `C:\opt\openvino` — adjust if yours differs:

```powershell
& "C:\opt\openvino\setupvars.ps1"
$env:Path = "$env:INTEL_OPENVINO_DIR\runtime\bin\intel64\Release;$env:INTEL_OPENVINO_DIR\runtime\3rdparty\tbb\bin;$env:Path"
```

Golden 64-token greedy parity vs `tools/qwen36_surgery/golden/qwen36_parity_64.json` (`Resolve-Path` makes it absolute — `cargo test` runs from the crate dir, not the workspace root):

```powershell
$env:QWEN36_SHARDS=(Resolve-Path ./qwen36-shards-2stage).Path; cargo test -p cascadia-engine-openvino --features openvino -- --ignored qwen36_greedy_parity
```

Serve the in-process stage chain (qwen36-moe loads all stages of the tree in one process):

```powershell
cargo run -p cascadia --release --features openvino -- run ./qwen36-shards-2stage --engine qwen36-moe --device CPU --api :8000
```

Assert a correct answer in a second PowerShell window (`enable_thinking:false` for a direct reply; without it the model emits a chain-of-thought first). Expect Paris; the 2-stage tree exercises rank0->rank1 forwarding in-process:

```powershell
(Invoke-RestMethod -Uri http://127.0.0.1:8000/v1/chat/completions -Method Post -ContentType 'application/json' -Body '{"model":"qwen36","messages":[{"role":"user","content":"Capital of France? One word."}],"max_tokens":16,"enable_thinking":false}').choices[0].message.content
```

Optional deeper checks (dev probes — `probe_engine_parity.py` emits the whole-model reference, `probe_chain_vs_full_prompt.py` is the chain-vs-full comparator; the cargo test above is the pass/fail verdict):

```powershell
python tools/qwen36_surgery/probe_engine_parity.py
```

### C — cross-node fleet + concurrency (Windows AI-PCs)

Owned by cascadia-enterprise#36: `scripts/qwen36/{setup,demo}.sh` bring up the N-stage chain on the fleet and the demo runs `prod_infer` (asserted correctness) + `prod_burst` (concurrency). Same engine code as this PR — **no separate fleet run needed here.**

## Known limitations / follow-ups
- **MoE multi-GPU fidelity**: f16 cross-stage hand-offs perturb the residual stream and can flip the 256-expert top-8 router (a discrete top-k decision — a cliff, not a slope). **2-stage and 3-stage are clean; 4-stage drifts** (garbles even short answers). **Validated deployment envelope is ≤3 stages.** Deterministic-kernel / gate-f32 OV config is the follow-up, not this PR.
- bf16 HF-reference parity not claimable on 32 GB fleet nodes (stated in the prompt-set gate).

Closes the community half of #77.

### CODE REVIEW

**Review by 7i7o — APPROVED** [review]
tACK 👍
